lib/rust/ensogl/core/src/data/dirty.rs

```diff
@@ -205,181 +270,303 @@ where Arg<T>: Display
+/// API but without requiring `self` reference to be mutable.
+#[derive(Derivative, CloneRef, From, Deref)]
+#[derivative(Debug(bound = "T:Debug"))]
+#[derivative(Clone(bound = ""))]
+#[repr(transparent)]
+pub struct SharedFlag<T, OnMut> {
+    rc: Rc<RefCellFlag<T, OnMut>>,
+}
+
+
+// === API ===
+
+impl<T: Default, OnMut> SharedFlag<T, OnMut> {
+    /// Constructor.
+    pub fn new(on_set: OnMut) -> Self {
+        Self { rc: Rc::new(RefCellFlag::new(on_set)) }
     }
 }
 
 
 // === Arg ===
 
-impl<T: HasArg, OnMut> HasArg for SharedDirtyFlag<T, OnMut> {
+impl<T: HasArg, OnMut> HasArg for SharedFlag<T, OnMut> {
     type Arg = Arg<T>;
 }
 
 
 // === Global Operations ===
 
-impl<T: HasUnsetAll, OnMut> SharedHasUnsetAll for SharedDirtyFlag<T, OnMut> {
+impl<T: HasUnsetAll, OnMut> SharedHasUnsetAll for SharedFlag<T, OnMut> {
     fn unset_all(&self) {
-        self.rc.borrow_mut().unset_all()
+        self.rc.unset_all()
     }
 }
 
-impl<T: HasCheckAll, OnMut> HasCheckAll for SharedDirtyFlag<T, OnMut> {
+impl<T: HasCheckAll, OnMut> HasCheckAll for SharedFlag<T, OnMut> {
     fn check_all(&self) -> bool {
-        self.rc.borrow().check_all()
+        self.rc.check_all()
     }
 }
 
 // === Check ===
 
-impl<T: DirtyFlagOps0, OnMut> HasCheck0 for SharedDirtyFlag<T, OnMut> {
+impl<T: FlagOps0, OnMut> HasCheck0 for SharedFlag<T, OnMut> {
     fn check(&self) -> bool {
-        self.rc.borrow().check()
+        self.rc.check()
     }
 }
 
-impl<T: DirtyFlagOps1, OnMut> HasCheck1 for SharedDirtyFlag<T, OnMut> {
+impl<T: FlagOps1, OnMut> HasCheck1 for SharedFlag<T, OnMut> {
     fn check(&self, arg: &Arg<T>) -> bool {
-        self.rc.borrow().check(arg)
+        self.rc.check(arg)
     }
 }
 
 // === Set ===
 
-impl<T: DirtyFlagOps0, OnMut: FnMut0> SharedHasSet0 for SharedDirtyFlag<T, OnMut> {
+impl<T: FlagOps0, OnMut: FnMut0> SharedHasSet0 for SharedFlag<T, OnMut> {
     fn set(&self) {
-        self.rc.borrow_mut().set()
+        self.rc.set()
     }
 }
 
-impl<T: DirtyFlagOps1, OnMut: FnMut0> SharedHasSet1 for SharedDirtyFlag<T, OnMut> {
+impl<T: FlagOps1, OnMut: FnMut0> SharedHasSet1 for SharedFlag<T, OnMut> {
     fn set(&self, arg: Arg<T>) {
-        self.rc.borrow_mut().set(arg)
+        self.rc.set(arg)
     }
 }
 
 // === Unset ===
 
-impl<T: HasUnset0, OnMut> SharedHasUnset0 for SharedDirtyFlag<T, OnMut> {
+impl<T: HasUnset0, OnMut> SharedHasUnset0 for SharedFlag<T, OnMut> {
     fn unset(&self) {
-        self.rc.borrow_mut().unset()
+        self.rc.unset()
     }
 }
 
-impl<T: HasUnset1, OnMut> SharedHasUnset1 for SharedDirtyFlag<T, OnMut>
+impl<T: HasUnset1, OnMut> SharedHasUnset1 for SharedFlag<T, OnMut>
 where Arg<T>: Display
 {
     fn unset(&self, arg: &Self::Arg) {
-        self.rc.borrow_mut().unset(arg)
+        self.rc.unset(arg)
     }
 }
 
 
-// === Weak References ===
 
+// ======================
+// === WeakSharedFlag ===
+// ======================
+
+/// A weak version of [`RefCellFlag`].
```

Comment:
```suggestion
/// A weak version of [`SharedFlag`].
```
This is the type having weak version, isn't it?